src/converters/app-converter.ts

```diff
@@ -0,0 +1,397 @@
+import * as pulumi from '@pulumi/pulumi';
+import { AssemblyManifestReader, StackManifest } from '../assembly';
+import { ConstructInfo, GraphBuilder } from '../graph';
+import { StackComponentResource, lift, Mapping } from '../types';
+import { ArtifactConverter, FileAssetManifestConverter } from './artifact-converter';
+import { CdkConstruct, ResourceMapping } from '../interop';
+import { debug } from '@pulumi/pulumi/log';
+import {
+    cidr,
+    getAccountId,
+    getAzs,
+    getRegion,
+    getSsmParameterList,
+    getSsmParameterString,
+    getUrlSuffix,
+} from '@pulumi/aws-native';
+import { mapToAwsResource } from '../aws-resource-mappings';
+import { attributePropertyName, mapToCfnResource } from '../cfn-resource-mappings';
+import { CloudFormationResource, getDependsOn } from '../cfn';
+import { OutputMap, OutputRepr } from '../output-map';
+import { parseSub } from '../sub';
+import { getPartition } from '@pulumi/aws-native/getPartition';
+
+/**
+ * AppConverter will convert all CDK resources into Pulumi resources.
+ */
+export class AppConverter {
+    // Map of stack artifactId to StackConverter
+    public readonly stacks = new Map<string, StackConverter>();
+
+    public readonly manifestReader: AssemblyManifestReader;
+
+    constructor(readonly host: StackComponentResource) {
+        this.manifestReader = AssemblyManifestReader.fromDirectory(host.assemblyDir);
+    }
+
+    convert() {
+        for (const stackManifest of this.manifestReader.stackManifests) {
+            const stackConverter = new StackConverter(this.host, stackManifest);
+            this.stacks.set(stackManifest.id, stackConverter);
+            this.convertStackManifest(stackManifest);
+        }
+    }
+
+    private convertStackManifest(artifact: StackManifest): void {
+        const dependencies = new Set<ArtifactConverter>();
+        for (const file of artifact.fileAssets) {
+            const converter = new FileAssetManifestConverter(this.host, file);
+            converter.convert();
+            dependencies.add(converter);
+        }
+
+        // TODO add docker asset converter
+        // for (const image of artifact.dockerAssets) {
+        // }
+
+        const stackConverter = this.stacks.get(artifact.id);
+        if (!stackConverter) {
+            throw new Error(`missing CDK Stack for artifact ${artifact.id}`);
+        }
+        stackConverter.convert(dependencies);
+    }
+}
+
+/**
+ * StackConverter converts all of the resources in a CDK stack to Pulumi resources
+ */
+export class StackConverter extends ArtifactConverter {
+    readonly parameters = new Map<string, any>();
+    readonly resources = new Map<string, Mapping<pulumi.Resource>>();
+    readonly constructs = new Map<ConstructInfo, pulumi.Resource>();
+    stackResource!: CdkConstruct;
```

Comment:
I'm always a little wary of mutable state. This starts out as undefined, right? Then it's set in convert. Is it ever accessed before it's set? Would it be easier to read and guarantee sequencing works out if this wasn't in the state at all? 

Also curious what does "!" do in TypeScript, reading up on this. 

GPT-4o:

```
In TypeScript, the =!= after a variable name (e.g., =stackResource!=) is called the definite assignment assertion. It
tells the TypeScript compiler that you are confident the property will be assigned a value before it is accessed, even
if the compiler cannot determine this based on its normal flow analysis.
```

Sounds like a possible footgun later on, but perhaps not worth rewriting now if we're confident.